src/agent/rustjail/src/container.rs

```diff
@@ -245,12 +255,53 @@ pub struct SyncPC {
 pub trait Container: BaseContainer {
     //	fn checkpoint(&self, opts: &CriuOpts) -> Result<()>;
     //	fn restore(&self, p: &Process, opts: &CriuOpts) -> Result<()>;
-    fn pause(&self) -> Result<()>;
-    fn resume(&self) -> Result<()>;
+    fn pause(&mut self) -> Result<()>;
+    fn resume(&mut self) -> Result<()>;
     //	fn notify_oom(&self) -> Result<(Sender, Receiver)>;
     //	fn notify_memory_pressure(&self, lvl: PressureLevel) -> Result<(Sender, Receiver)>;
 }
 
+impl Container for LinuxContainer {
+    fn pause(&mut self) -> Result<()> {
+        let status = self.status();
+        if status == Status::RUNNING || status == Status::CREATED {
+            if self.cgroup_manager.is_some() {
+                self.cgroup_manager
+                    .as_ref()
+                    .unwrap()
+                    .freeze(fscgroup::FROZEN)?;
+            }
+
+            self.status = Status::PAUSED;
+            return Ok(());
+        }
+
+        Err(ErrorKind::ErrorCode(format!(
+            "failed to pause container: status is: {:?}",
+            status
+        ))
+        .into())
+    }
+
+    fn resume(&mut self) -> Result<()> {
```

Comment:
Nit: This function isn't the "opposite" of `pause` in the sense that `resume()` checks for the starting state and errors immediately if not correct. However, `pause()` handles that right at the end.